refactor(ast): reduce `#[cfg_attr]` boilerplate in AST type defs

overlookmotel opened the pull request:
Remove a ton of `#[cfg_attr(feature = "serialize", serde(...))]` boilerplate from AST type definitions.

Before: `#[cfg_attr(feature = "serialize", serde(flatten))]`
After: `#[serde(flatten)]`

This is a reprise of #2669, which was later reverted, but this time doing it using our existing zero-cost `#[ast]` dummy macro attr, so no compile time penalty this time around.

This makes no difference to either runtime or compile time behavior, purely removes the `cfg_attr` boilerplate and makes the code easier to read.

overlookmotel commented:
* **#4375** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4375?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

rzvxa commented on crates/oxc_ast/src/ast/jsx.rs:
```diff
@@ -23,9 +23,9 @@ use super::{inherit_variants, js::*, literal::*, ts::*};
 #[ast(visit)]
 #[derive(Debug, Hash)]
 #[cfg_attr(feature = "serialize", derive(Serialize, Tsify))]
```
Can't we also put these in `#[ast]`? Similar to how we are adding ast derive.

rzvxa replied:
To clarify I'm talking about this line:

```rust
#[cfg_attr(feature = "serialize", derive(Serialize, Tsify))]
```

overlookmotel replied:
You mean add `#[cfg_attr(feature = "serialize", derive(Serialize, Tsify))]` in `#[ast]` macro so we can remove it from type defs?

If that's what you mean, I think this would be a more "controversial" change, because it hides away things from the reader ("How does this type get serialized? It doesn't implement `Serialize`!"). In comparison, the changes in this PR don't really do that - they just strip out some largely meaningless syntax noise. Pretty sure Boshen would be against going so far as hiding `#[derive]`s.

overlookmotel committed:
refactor(ast): reduce `#[cfg_attr]` boilerplate in AST type defs (#4375)

Remove a ton of `#[cfg_attr(feature = "serialize", serde(...))]` boilerplate from AST type definitions.

Before: `#[cfg_attr(feature = "serialize", serde(flatten))]`
After: `#[serde(flatten)]`

This is a reprise of #2669, which was later reverted, but this time doing it using our existing zero-cost `#[ast]` dummy macro attr, so no compile time penalty this time around.

This makes no difference to either runtime or compile time behavior, purely removes the `cfg_attr` boilerplate and makes the code easier to read.
